crates/ruff_db/src/vendored.rs

```
@@ -0,0 +1,379 @@
+        let normalized = normalize_vendored_path(path);
+        let inner_locked = self.inner.lock();
+        let mut archive = inner_locked.borrow_mut();
+        if let Ok(zip_file) = archive.lookup_path(&normalized) {
+            if zip_file.is_dir() {
+                return Some(FileType::Directory);
+            }
+            return Some(FileType::File);
+        }
+        if let Ok(zip_file) = archive.lookup_path(&normalized.with_trailing_slash()) {
+            if zip_file.is_dir() {
+                return Some(FileType::Directory);
+            }
+            return Some(FileType::File);
+        }
+        None
+    }
+
+    pub fn exists(&self, path: &VendoredPath) -> bool {
+        let normalized = normalize_vendored_path(path);
+        let inner_locked = self.inner.lock();
+        let mut archive = inner_locked.borrow_mut();
+        archive.lookup_path(&normalized).is_ok()
+            || archive
+                .lookup_path(&normalized.with_trailing_slash())
+                .is_ok()
+    }
+
+    /// Return the crc32 hash of the file
+    pub fn file_hash(&self, path: &VendoredPath) -> Option<u32> {
+        let normalized = normalize_vendored_path(path);
+        let inner_locked = self.inner.lock();
+        let mut archive = inner_locked.borrow_mut();
+        if let Ok(zip_file) = archive.lookup_path(&normalized) {
+            return Some(zip_file.crc32());
+        }
+        archive
+            .lookup_path(&normalized.with_trailing_slash())
+            .map(|zip_file| zip_file.crc32())
+            .ok()
+    }
+
+    /// Read the entire contents of the path into a string
+    pub fn read(&self, path: &VendoredPath) -> Option<String> {
+        let inner_locked = self.inner.lock();
+        let mut archive = inner_locked.borrow_mut();
+        let mut zip_file = archive.lookup_path(&normalize_vendored_path(path)).ok()?;
+        let mut buffer = String::new();
+        zip_file.read_to_string(&mut buffer).ok()?;
```

What are the reasons why `read_to_string` can fail? I'm not sure if just silently returning `None` in that case is the best solution. Because the file does exist, it just can't be read for obscure reasons. We may want to change this method to return a result instead.

---

It returns an `Err` if the data that it is trying to read into the `String` buffer turns out not to be valid UTF-8. I agree that we shouldn't silently return `None` here if that happens.